src/compiler/transformers/es2017.ts

```diff
@@ -620,6 +664,89 @@ namespace ts {
         }
     }
 
+    /** Creates a variable named `_superProps` with accessor properties for the given property names. */
+    export function createSuperAccessVariableStatement(resolver: EmitResolver, node: FunctionLikeDeclaration, names: UnderscoreEscapedMap<true>) {
+        // Create a variable declaration with a getter/setter (if binding) definition for each name:
+        //   const _superProps = Object.create(null, { x: { get: () => super.x, set: (v) => super.x = v }, ... });
+        const hasBinding = (resolver.getNodeCheckFlags(node) & NodeCheckFlags.AsyncMethodWithSuperBinding) !== 0;

```

Comment:
Note that, to track whether the name is used for reading or writing would require a case in `visitor` for every assignment-capable node, e.g.:

```ts
function visitor(node: Node) {
  switch (node.kind) {
    …
    case SyntaxKind.BinaryExpression:
      return visitBinaryExpression(<BinaryExpression>node);
  }
  …
}

function assignmentTargetVisitor(node: Node) {
  // visitor that descends through assignment targets:
  // - left side of any assignment expression
  // - array literal (destructuring assignment)
  // - object literal (destructuring assignment)
  // - property assignment (destructuring assignment)
  // - etc.
  // if the node is a SuperProperty, we can visit it knowing that we are in an assignment position.
}

function visitBinaryExpression(node: BinaryExpression) {
  // only need to special case if the assignment target contains `super`
  if (isAssignmentExpression(node) && node.left.transformFlags & TransformFlags.ContainsSuper) {
    return updateBinary(
      visitNode(node.left, assignmentTargetVisitor, isExpression), // visit left side as assignment.
      node.operatorToken,
      visitNode(node.right, visitor, isExpression));
  }
  return visitEachChild(node, visitor, context);
}
```

Reply:
Agreed, with the code as is, we're over-generating setters in this situation.

I'm not sure if it's worth the added complexity though. As the original comment in checker states, we expect async methods accessing `super` to be rather uncommon, and I'd expect async methods accessing multiple different `super` properties to be even less common (I think most methods only ever call a single `super` method, their own parent, otherwise just dispatch via this), and then on top of that, async methods that are assigning into some super properties but not others where the over-generation of setters would kick in, to be very rare. These are all just guesses based on what code I've seen in the wild, but I think very few users will ever run into this situation.

On the flip side, we'd need to special case all assignments here and I think would also need to duplicate the code given the structure between `es2017.ts` and `esnext.ts` here.

I'm happy to make the change though if you think it's warranted/carrying its own weight in complexity.

Reply:
Probably not for now, but we may want to consider it in the future.